Cesium3DTilesSelection/src/Tile.cpp

```diff
@@ -297,65 +302,81 @@ void Tile::loadContent() {
-                  pContent->model->generateMissingNormalsSmooth();
-                }
-
-                if (pPrepareRendererResources) {
-                  CESIUM_TRACE("prepareInLoadThread");
-                  pRendererResources =
-                      pPrepareRendererResources->prepareInLoadThread(
-                          pContent->model.value(),
-                          transform);
-                }
-              }
-            }
-
-            LoadResult result;
-            result.state = LoadState::ContentLoaded;
-            result.pContent = std::move(pContent);
-            result.pRendererResources = pRendererResources;
-
-            return result;
+            loadInput.asyncSystem = std::move(asyncSystem);
+            loadInput.pLogger = std::move(pLogger);
+            loadInput.pAssetAccessor = std::move(pAssetAccessor);
+            loadInput.pRequest = std::move(pRequest);
+
+            return TileContentFactory::createContent(loadInput)
+                // Forward status code to the load result.
+                .thenInWorkerThread(
+                    [statusCode = pResponse->statusCode(),
+                     loadInput = std::move(loadInput),
+                     gltfUpAxis,
+                     projections = std::move(projections),
+                     generateMissingNormalsSmooth,
+                     pPrepareRendererResources =
+                         std::move(pPrepareRendererResources)](
+                        std::unique_ptr<TileContentLoadResult>&&
+                            pContent) mutable {
+                      void* pRendererResources = nullptr;
+
+                      if (pContent) {
+                        pContent->httpStatusCode = statusCode;
+                        if (statusCode < 200 || statusCode >= 300) {
+                          return LoadResult{
+                              LoadState::FailedTemporarily,
+                              std::move(pContent),
+                              nullptr};
+                        }
+
+                        if (pContent->model) {
+
+                          CesiumGltf::Model& model = pContent->model.value();
+
+                          // TODO The `extras` are currently the only way to
+                          // pass arbitrary information to the consumer, so the
+                          // up-axis is stored here:
+                          model.extras["gltfUpAxis"] = gltfUpAxis;
+
+                          Tile::generateTextureCoordinates(
+                              model,
+                              loadInput.tileTransform,
+                              // Would it be better to use the content bounding
+                              // volume, if it exists?
+                              // What about
+                              // TileContentLoadResult::updatedBoundingVolume?
+                              loadInput.tileBoundingVolume,
+                              projections);
```

Comment:
@kring I think everything dubious has been fixed now! What are your thoughts on this last thing, does it make a difference which bounding volume is used for the texture coordinates generation?

Reply:
That's a really interesting question!

I think we need to use the original tile bounding volume for texture coordinate generation, not the content bounding volume. That's because we sometimes expect to be able to use parent tile raster overlay textures when rendering child tiles. And we can't do that reliably if we use a bounding volume that (pretty much by definition) doesn't cover parts of the child tiles. At the very least, the raster overlay tile selection algorithm would get more complicated. Currently if all the overlay tiles that are attached to a child tile fail to load, we use the parent texture. But if the parent texture coordinate space doesn't cover the child, we couldn't do that anymore. We'd have to load additional parent-level raster overlay tiles for use with the child. So I think we should keep this simple until we identify a tileset or use-case where the extra complexity provides a compelling benefit.

As for the `updatedBoundingVolume`, it's unfortunately not totally clear (either in my head or in the doc) whether this is meant to bound the content or the content plus all child content. In the case where it's used - quantized mesh terrain - the two are the same so the question didn't really come up. But `Tile::update` uses it to update the tile bounding volume (not the content bounding volume), so I guess that's something like an authoritative answer. So, since that field is an "improved" version of the tile bounding volume, I'd say that implies we should use it to generate texture coordinates if it's available.